hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,343 @@
+    this.networkTopology = networkTopology;
+    this.metrics = metrics;
+  }
+
+  /**
+   * Called by SCM to choose datanodes.
+   *
+   *
+   * @param excludedNodes - list of the datanodes to exclude.
+   * @param favoredNodes - list of nodes preferred. This is a hint to the
+   *                     allocator, whether the favored nodes will be used
+   *                     depends on whether the nodes meets the allocator's
+   *                     requirement.
+   * @param nodesRequired - number of datanodes required.
+   * @param dataSizeRequired - size required for the container.
+   * @param metadataSizeRequired - size required for Ratis metadata.
+   * @return List of datanodes.
+   * @throws SCMException  SCMException
+   */
+  @Override
+  public List<DatanodeDetails> chooseDatanodes(
+      List<DatanodeDetails> excludedNodes, List<DatanodeDetails> favoredNodes,
+      int nodesRequired, long metadataSizeRequired, long dataSizeRequired)
+      throws SCMException {
+    Preconditions.checkArgument(nodesRequired > 0);
+    metrics.incrDatanodeRequestCount(nodesRequired);
+    int datanodeCount = networkTopology.getNumOfLeafNode(NetConstants.ROOT);
+    int excludedNodesCount = excludedNodes == null ? 0 : excludedNodes.size();
+    if (datanodeCount < nodesRequired + excludedNodesCount) {
+      throw new SCMException("No enough datanodes to choose. " +
+          "TotalNode = " + datanodeCount +
+          " RequiredNode = " + nodesRequired +
+          " ExcludedNode = " + excludedNodesCount, null);
+    }
+
+    List<DatanodeDetails> mutableFavoredNodes = new ArrayList<>();
+    if (favoredNodes != null) {
+      // Generate mutableFavoredNodes, only stores valid favoredNodes
+      for (DatanodeDetails datanodeDetails : favoredNodes) {
+        if (isValidNode(datanodeDetails, metadataSizeRequired,
+            dataSizeRequired)) {
+          mutableFavoredNodes.add(datanodeDetails);
+        }
+      }
+    }
+    if (excludedNodes != null) {
+      mutableFavoredNodes.removeAll(excludedNodes);
+    }
+
+    // For excluded nodes, we sort their racks at rear
+    List<Node> racks = getAllRacks();
+    if (excludedNodes != null && excludedNodes.size() > 0) {
+      racks = sortRackWithExcludedNodes(racks, excludedNodes);
+    }
+
+    List<Node> toChooseRack = new LinkedList<>(racks);
+    List<Node> chosenNodes = new ArrayList<>();
+    List<Node> unavailableNodes = new ArrayList<>();
+    if (excludedNodes != null) {
+      unavailableNodes.addAll(excludedNodes);
+    }
+
+    // If the result doesn't change after retryCount, we return with exception
+    int retryCount = 0;
+    while (nodesRequired > 0) {
+      if (retryCount > MAX_RETRY) {
+        throw new SCMException("No satisfied datanode to meet the" +
+            " excludedNodes and affinityNode constrains.", null);
+      }
+      int chosenListSize = chosenNodes.size();
+
+      if (toChooseRack.size() == 0) {
+        toChooseRack.addAll(racks);
```

Comment:
This will be executed on a retry pass. I have a couple of questions:

1. Do we need to remove all the racks in chosenNodes here?

2. When do we expect the loop to run again for a retry? Is this related to fallback, or do we just retry a few times if we don't find enough nodes on unique racks?

Reply:
> Do we need to remove all the racks in chosenNodes here?

I think we don't need to remove the content of chosenNodes. The size of chosenNodes could be used the check if it's a retry or not.

> When do we expect the loop to run again for a retry? Is this related to fallback, or do we just retry a few times if we don't find enough nodes on unique racks?

If the size of chosenNodes doesn't change, we treat it as a retry, and if the size doesn't change for RETRY_MAX times, then we return with exception.

Reply:
What I meant is, do we need to remove all the racks already used in choosenNodes from toChooseRack after we all all the racks to it, otherwise we might use the same rack twice? There are two potential problems here - one is that we did not find all the nodes on the first pass over all racks, perhaps as there were not enough nodes with free space selected randomly from a certain rack. However if we tried again, we might get some good nodes and obtain the correct number of racks.

The other problem is that there might not be enough racks to meet the goal, and therefore we cannot remove them from the list.

I **think** we might see some corner cases where we don't select enough racks even though we should be able too.

Reply:
I think if all nodes in each rack are available, the outer loop would be used mostly for situations like "10 replicas and 2 racks" (numReplicas > numRacks), so that for each loop we can get 2 nodes.

For the first problem, I was thinking we can add a list to store the racks not used in this loop, and put them at the head of toChooseRack so that they will be easier chosen.
Another thing is we can enlarge the RETRY_COUNT of inner loop, so that the first problem is more likely to be eliminated.